File: leetcode/904-leaf-similar-trees/leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void __leafNodes(TreeNode* node,vector<int> &leaf){
        if(node==nullptr){
            return;
        }
        __leafNodes(node->left,leaf);
        if(node->left==nullptr && node->right==nullptr){
            leaf.push_back(node->val);
        }
        __leafNodes(node->right,leaf);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> leaf1;
        vector<int> leaf2;
        __leafNodes(root1,leaf1);
        __leafNodes(root2,leaf2);
        if(leaf1.size()!=leaf2.size()){
            return false;
        }
        for(int i=0;i<leaf1.size();i++){
            if(leaf1[i]!=leaf2[i]){
                return false;
            }
        }
        return true;
    }
};

File: leetcode/904-leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def leafSimilar(self, root1, root2):
        """
        :type root1: Optional[TreeNode]
        :type root2: Optional[TreeNode]
        :rtype: bool
        """
        list1 = []
        list2 = []
        def helper(node,list):
            if not node:
                return
            if not node.left and not node.right:
                list.append(node.val)
                return
            
            helper(node.left,list)
            helper(node.right,list)

        helper(root1,list1)
        helper(root2,list2)
        if len(list1) != len(list2):
            return False
        
        for i in range(len(list1)):
            if list1[i] != list2[i]:
                return False
        return True
        